tests/test_main.py

```diff
@@ -22,6 +22,11 @@ def test_set_key_no_file(tmp_path):
     assert os.path.exists(nx_file)
 
 
+def test_set_key_invalid_file():
+    with pytest.raises(TypeError):
+        dotenv.set_key(None, "foo", "bar")  # type: ignore
```

Comment:
I'm not sure we need (or should) test this. `None` is not supposed to be used for the first argument, so whatever happens in that case is not really the responsibility of the library, in my opinion. Same if you try with integers or other unsupported types.

However, there's one test that we could add, that would fail in the `main` branch and succeed in this pull request. Something like this:

```python
def test_set_key_unauthorized_file():
    with pytest.raises(PermissionError):
        dotenv.set_key("/root/unauthorized", "a", "x")
```

What do you think?

Reply:
Touching anything in `/root/` in a test terrifies me.  Sure, your test runner shouldn't be a sudoer, but even then this raises questions about the platform your tests are running on or whether the tests should still pass if run as a sudoer.  

I do think the test as written is nonetheless realistic:  it's how I discovered the out-of-scope error in python-dotenv.  Some upstream code paths led to the file being `None`, and while I would have expected this to be rejected as an invalid argument when `set_key` was called, it was not and instead I received the error internal to python-dotenv, #413.  I do understand your desire to avoid having to use the `#type: ignore`, and I propose this instead:

```
def test_set_key_unauthorized_file(dotenv_file):
    os.chmod(dotenv_file, 0o000)

    with pytest.raises(PermissionError):
        dotenv.set_key(dotenv_file, "a", "x")
```

Reply:
This is much better :+1: